src/noit_metric_director.c

```diff
@@ -280,47 +413,70 @@ noit_metric_director_flush(eventer_t e) {
 
 static void
 distribute_metric(noit_metric_message_t *message) {
-  void *vhash, *vinterests;
-  caql_cnt_t *interests = NULL;
-  char uuid_str[UUID_STR_LEN + 1];
-  mtev_uuid_unparse_lower(message->id.id, uuid_str);
-  if(mtev_hash_retrieve(&id_level, (const char *) &message->id.id, UUID_SIZE,
-      &vhash)) {
-    if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
-        message->id.name_len_with_tags, &vinterests)) {
-      interests = vinterests;
-      distribute_message_with_interests(interests, message);
+  int i;
+  static uuid_t uuid_zero = { 0 };
+  void *vhash, **vinterests;
+  interest_cnt_t has_interests = 0;
+  interest_cnt_t interests[nthreads];
+  memset(interests, 0, sizeof(interests));
+
+
+  mtev_memory_begin();
+  /* First we process interests for specific metrics: uuid-metric|ST[tags...] */
+  for(i=0; i<2; i++) {
+    /* First time through we use the message's ID, second time we use the NULL UUID
+     * which is used to "match any" check.
+     */
+    if(mtev_hash_retrieve(&id_level, i ? (const char *)&uuid_zero : (const char *)&message->id.id, UUID_SIZE,
+        &vhash)) {
+      if(mtev_hash_retrieve((mtev_hash_table *) vhash, message->id.name,
+                            message->id.name_len_with_tags, (void **)&vinterests)) {
+        memcpy(interests, *vinterests, sizeof(interests)); /*  we access this exactly once, no need to ck load or tmp variable */
```

Comment:
For this line and 439, the code is correct... but the comment is not.  What makes it safe is 

1. SMR
1. That we'll see either the old pointer of the array or the new one.

In other words, we're not using a fence to enforce visibility but we're OK with just atomicity.